packages/core/src/common/message-service-protocol.ts

```diff
@@ -16,25 +16,64 @@
 
 import { injectable, inject } from 'inversify';
 import { ILogger } from './logger';
+import { CancellationToken } from './cancellation';
 
 export const messageServicePath = '/services/messageService';
 
 export enum MessageType {
     Error = 1,
     Warning = 2,
     Info = 3,
-    Log = 4
+    Log = 4,
+    Progress = 5
 }
 
 export interface Message {
-    type: MessageType;
-    text: string;
-    actions?: string[];
-    options?: MessageOptions;
+    readonly type?: MessageType;
+    readonly text: string;
+    readonly actions?: string[];
+    readonly options?: MessageOptions;
+}
+
+export interface ProgressMessage extends Message {
+    readonly type?: MessageType.Progress;
+    readonly options?: ProgressMessageOptions;
+}
+export namespace ProgressMessage {
+    export const Cancel = 'Cancel';
+    export function isCancelable(message: ProgressMessage): boolean {
+        return !message.options
+            || message.options.cancelable === undefined
+            || message.options.cancelable === true;
+    }
 }
 
 export interface MessageOptions {
-    timeout?: number;
+    readonly timeout?: number;
+}
+
+export interface ProgressMessageOptions extends MessageOptions {
+    /**
+     * Default: `true`
+     */
+    readonly cancelable?: boolean;
+}
+
+export interface Progress {
+    readonly id: string;
+    readonly report: (update: ProgressUpdate) => void;
+    readonly cancel: () => void;
+    readonly result: Promise<string | undefined>;
+}
+
+export interface ProgressUpdate {
+    readonly message?: string;
```

Comment:
couldn't that be mandatory?

Reply:
I need to stress the BE use case again: if you send the message only once at the beginning, and a client joins later on (e.g. because of browser refresh,) then there wouldn't be any message to render on next progress report arriving on the FE.

for the common understanding of the API. can we agree, that the report update always provides the complete info what to render?

Reply:
I agree, I'll rework the API a bit to always receive complete info.

Reply:
Just to be sure that we are on the same wave. We have complete info to render on update step:
https://github.com/theia-ide/theia/blob/38b7606f24aa80dca13c9ec44b632c47b2582257/packages/core/src/common/message-service-protocol.ts#L111
We have mandatory `message: ProgressMessage` argument and it would be rendered in case when client re-connection.

Reply:
looks good.